router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+							)
+
+							// If the error is not recoverable, return it and abort the poller
+							return fetchError.Err
+						}
+					} else {
+						p.logger.Error("fetch error", zap.Error(fetchError.Err), zap.String("topic", fetchError.Topic))
+					}
+				}
+			}
+
+			iter := fetches.RecordIter()
+			for !iter.Done() {
+				r := iter.Next()
+
+				rec := &record{
+					topic: r.Topic,
+					data:  r.Value,
+				}
+
+				p.updateTriggers(rec)
+			}
+		}
+	}
+}
+
+// updateTriggers updates all subscriptions that are interested in the topic with the new data.
+func (p *kafkaPubSub) updateTriggers(rec *record) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	// Notify all subscriptions that are interested in the topic
+	if sub, ok := p.subscriptions[rec.topic]; ok {
+
+		for _, s := range sub {
+			// If the subscription is still active, update it
+			if s.context.Err() == nil {
+				p.logger.Debug("subscription update", zap.String("topic", rec.topic), zap.ByteString("data", rec.data))
+				s.updater.Update(rec.data)
+			}
+		}
+	}
+}
+
+// Subscribe subscribes to the given topics and updates the subscription updater.
+// The engine already deduplicates subscriptions with the same topics, stream configuration, extensions, headers, etc.
+func (p *kafkaPubSub) Subscribe(ctx context.Context, event pubsub_datasource.KafkaSubscriptionEventConfiguration, updater resolve.SubscriptionUpdater) error {
+
+	p.logger.Debug("subscribe",
+		zap.Strings("topics", event.Topics),
+		zap.String("providerID", event.ProviderID),
+	)
+
+	sub := &subscription{
+		topics:  event.Topics,
+		updater: updater,
+		context: ctx,
+	}
+
+	// Every subscription needs to be tracked individually
+	p.addSubscription(sub)
+
+	for _, t := range event.Topics {
+
+		topic := t
+
+		client, created, err := p.createClient(topic)
+		if err != nil {
+			p.logger.Error("failed to create client", zap.Error(err), zap.String("topic", topic))
+			return err
+		}
+
+		// Only continue and start the poller if the client was created for the first time
+		if !created {
+			continue
+		}
+
+		go func() {
+			err := p.topicPoller(client)
+			if err != nil {
+				if !errors.Is(err, errClientClosed) {
+					p.logger.Error("poller error", zap.Error(err))
+				}
+				return
+			}
+		}()
+	}
+
+	go func() {
+
+		<-sub.context.Done()
+
+		p.logger.Debug("subscription canceled",
+			zap.String("providerID", event.ProviderID),
+		)
+
+		p.removeSubscription(sub, &event)
+	}()
+
+	return nil
+}
+
+// removeSubscription removes the subscription from the topics and closes the client if there are no more subscriptions for the topic.
+func (p *kafkaPubSub) removeSubscription(sub *subscription, event *pubsub_datasource.KafkaSubscriptionEventConfiguration) {
+
+	// Remove the subscription from the topics
+	for _, topic := range event.Topics {
+
+		p.mu.RLock()
+		// Zero alloc due to slice reuse
+		tmp := p.subscriptions[topic][:0]
+		for _, s := range p.subscriptions[topic] {
```

Comment:
Have you considered using a map with sub as the key?

Reply:
How would it help me to manage subscriptions to topics?